### ISSUE TITLE
Tree-sitter compatibility with stable and beta

### ISSUE THREAD

**Issue by Aerijo** [atom/atom#17777]
A hurdle to writing Tree-sitter grammars for Atom (that I've hit at least) is versioning of everything. What I've gathered is that, at this point in time, 
- Stable Atom uses `tree-sitter-cli: "^0.12.5"` , and 
- Beta Atom uses `tree-sitter-cli: "^0.13.5"`

Just now, if I built the parser for one then it would not work in the other. Building it for beta caused stable to refuse to open the file at all; it just effectively ignored any attempt to do so. Building for stable gave a broken parse tree in beta.

(I already say build it locally with node version 8.9.3 to avoid other version issues besides dependencies.)

Is there a guide or best practice here? If I were writing a grammar for use in Atom, what should I do to make it compatible with whatever version of Atom the user has? I'm trying to write a beginner's guide for this, so any tips are appreciated.

The goal is to supplement the official docs with a "from scratch" guide for anyone who wants to make a grammar for use in Atom. It's started [here](https://gist.github.com/Aerijo/df27228d70c633e088b0591b8857eeef).

/cc @maxbrunsfeld ?

**Comment by maxbrunsfeld** [comment]
Yeah, sorry this has been a bumpy ride. I'm hopeful that soon, I'll stop having to make changes to Tree-sitter's parser ABI.

> What I've gathered is that, at this point in time, stable Atom uses tree-sitter-cli: "^0.12.5" , and beta Atom uses tree-sitter-cli: "^0.13.5"

Yeah, my approach for now is to use the npm modules' minor version numbers to indicate the Tree-sitter ABI version that they use. So for the `tree-sitter` and `tree-sitter-cli` modules, as well as every parser, `0.13.x` is the latest ABI version.

> Building it for beta caused stable to refuse to open the file at all

Yeah, that is due to an a bug in Atom that was fixed in https://github.com/atom/atom/pull/17424.

> Building for stable gave a broken parse tree in beta.

What did you see in this case? Hopefully you got an error in the console and it failed to load the grammar? That's the current expected behavior.

> The goal is to supplement the official docs with a "from scratch" guide for anyone who wants to make a grammar for use in Atom.

That would be so great! I don't have a great plan for dealing with this other than to stop changing the ABI version. Hopefully there will be no more changes for a while so that once Tree-sitter is enabled by default, people don't have to think about versioning.

For now, since this is an experimental feature, I would just advise building with the latest `tree-sitter-cli` and targeting beta only.

**Comment by maxbrunsfeld** [comment]
Also, in case you hadn't seen them, the existing docs for creating grammars are at http://tree-sitter.github.io/tree-sitter/creating-parsers. These docs aren't focused on Atom specifically though.

**Comment by Aerijo** [comment]
> Hopefully you got an error in the console and it failed to load the grammar? 

I got a couple of errors, but it did try to apply something to the file. Those stars kept multiplying too; there were none when I first checked the scope, then I checked other places and came back to find three. It also says it's using the grammar in the status bar (this is `v1.30.0-beta1`)

<img width="1279" alt="screen shot 2018-08-04 at 9 13 19 am" src="https://user-images.githubusercontent.com/22167388/43669512-eefb04aa-97c6-11e8-8180-e4f51b83dc8f.png">

> I'm hopeful that soon, I'll stop having to make changes to Tree-sitter's parser ABI.

Yeah, I figured things would settle down eventually. Would it still be an issue if you ever update the ABI when Tree-sitter becomes enabled by default? Or would that not need to happen to still make updates? 

And wouldn't there also be issues when the Nodejs version itself is ungraded? I noted that I build with node 8.9.3 installed because I ran into `NODE_MODULE_VERSION` errors a lot when starting out. 

I don't know how much of this can be automated by Atom, because I don't actually know much about the generated files (especially how the `src` files are supposed to be used to write `external` rules). Maybe the language package on Atom's side can specify specific tree-sitter parser versions for different Atom version ranges? It would mean updates can be safely released without breaking stuff for half the people.
